packages/store/addon/-private/system/model/model.js

```diff
@@ -1363,13 +1363,34 @@ if (DEBUG) {
     return isBasicDesc(instanceDesc) && lookupDescriptor(obj.constructor, keyName) === null;
   };
 
+  const INSTANCE_DEPRECATIONS = new WeakMap();
+  const DEPRECATED_LIFECYCLE_EVENT_METHODS = [
+    'becameError',
+    'becameInvalid',
+    'didCreate',
+    'didDelete',
+    'didLoad',
+    'didUpdate',
+    'ready',
+    'rolledBack',
+  ];
+
+  let lookupDeprecations = function lookupInstanceDeprecations(instance) {
+    let deprecations = INSTANCE_DEPRECATIONS.get(instance);
+
+    if (!deprecations) {
+      deprecations = new Set();
+      INSTANCE_DEPRECATIONS.set(instance, deprecations);
```

Comment:
I wonder if there is a way to avoid creating these data structures in memory if a user doesn't define any of these at all.  Perhaps a majority don't?  What do you think?

Reply:
I definitely agree that the majority a model classes won't be defining any of these methods. I'm not sure if there is a way to avoid creating these but I'm not sure it matters much since this code will only run in development. It's not clear from this diff but this is all wrapped inside of an `if (DEBUG)`